[flutter_tools] fix version tag `v` stripping and support old "dev" and new "pre" tags

peterlauri opened the pull request:
## Description

Correctly replacing 'v' in the version tag.

Supporting both "dev" and "pre" tag for **flutter version**.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/55421
Originates from https://github.com/flutter/flutter/pull/55385

## Tests

I added the following tests:

- Added missing test for switching to old "dev" tags.
- Added missing test for switching to new "pre" tags..

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

peterlauri committed:
fixes 55383

peterlauri committed:
fixes 55383

peterlauri committed:
Revert "fixes 55383"

This reverts commit 57948090b7116e451f716b6a2215965a3a8298a3.

peterlauri committed:
added test to prove that current implementation was wrong

peterlauri committed:
test case were fiddeling with wrong version string, fixed

peterlauri committed:
mimced formatting of other tests just to get going

peterlauri committed:
Merge branch 'master' into fix-flutter-version-dev-55383

peterlauri committed:
fixed regexp to support old "dev" and new "pre" tag format

peterlauri committed:
added test for "pre" type of tag

peterlauri committed:
Merge remote-tracking branch 'upstream/master' into fix-flutter-version-dev-55383

peterlauri commented:
@christopherfujino here is the new PR with tests and support for old "dev" and new "pre" tags.

peterlauri commented:
@christopherfujino this PR is up to date with upstream/master, so all should be good...

christopherfujino commented:
Sorry, I was busy today, I'll take a look at this on Monday though! Thanks!

peterlauri commented:
@christopherfujino cool, thanks...

christopherfujino commented:
Okay, sorry it took so long to review this, but I had to dive into why exactly this change was necessary.

This change is indeed correct, in that it will allow us to properly `flutter version 1.2.3-dev.0.0` AND the newer format `flutter version 1.2.3-0.0.pre` while accurately preserving the output "switching Flutter to version <version>". However this comes at the cost of forever having to implement the parsing logic for the old tags just for that output. Also, to truly make this correct, we would have to support the older `1.2.3-hotfix.1` and `1.2.3+hotfix.1` tags.

I think better would be to change the `flutter version` logic to not instantiate a new `FlutterVersion` after checking out the new version (https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/version.dart#L130), as this requires `FlutterVersion` to support parsing every type of version tag possible. Instead we could invoke git, and either print out the revision or use `git describe --match *.*.* --first-parent --tags`.

What do you think @zanderso @jonahwilliams @jmagman ?

christopherfujino reviewed (commented):
> Instead we could invoke git, and either print out the revision or use `git describe --match *.*.* --first-parent --tags`.

Or we could just print out the tag that we know we checked out.

jmagman commented:
We could have a `FlutterVersion` per format with a `-tryParse` and totally separate regexes, and first it tries the `PreFlutterVersion`, then `DevFlutterVersion`, then `HotfixFlutterVersion`...  This also will make it easy when, in the future, heaven forbid, we add a new variant.

christopherfujino commented:
> We could have a `FlutterVersion` per format with a `-tryParse` and totally separate regexes, and first it tries the `PreFlutterVersion`, then `DevFlutterVersion`, then `HotfixFlutterVersion`... This also will make it easy when, in the future, heaven forbid, we add a new variant.

All just for that one print statement?

jmagman commented:
> > We could have a `FlutterVersion` per format with a `-tryParse` and totally separate regexes, and first it tries the `PreFlutterVersion`, then `DevFlutterVersion`, then `HotfixFlutterVersion`... This also will make it easy when, in the future, heaven forbid, we add a new variant.
> 
> All just for that one print statement?

No, I meant more generically.  But I guess we're past that point.

christopherfujino commented:
> No, I meant more generically. But I guess we're past that point.

I don't think any other code would need to parse the old tags at this point.

zanderso commented:
I'd be reluctant to increase the complexity of the parsing here if the only reason to do it is to get `flutter version` to print the right string when switching versions, since that is a use-case we're trying to discourage/deprecate. Are there any other reasons to do it?

christopherfujino commented:
> I'd be reluctant to increase the complexity of the parsing here if the only reason to do it is to get `flutter version` to print the right string when switching versions, since that is a use-case we're trying to discourage/deprecate. Are there any other reasons to do it?

Not that I can think of. In fact, looking at the code again, I think it's fine to just print out the tag we checked out, because at the end of the command we invoke the doctor, which should use the parsing logic native to that version.

christopherfujino commented:
@peterlauri I wanted to make a PR into your feature branch, but I couldn't figure out how to do it from the GitHub web UI. Can you patch this diff into your PR https://github.com/flutter/flutter/pull/56093/commits/5658f5e42b301ed6fcd18bf118b75097a0df0962? (If you fetch upstream, I bet you can cherry pick the exact commit, I committed on top of your branch.)

It reverts the support for parsing the older version, and instead modifies the text that the `flutter version` tool prints out to not depend on parsing git tags.

peterlauri committed:
remove instantiating new FlutterVersion() in flutter version command. https://github.com/flutter/flutter/commit/5658f5e42b301ed6fcd18bf118b75097a0df0962

peterlauri commented:
I applied that patch to the PR.
https://github.com/flutter/flutter/pull/55602/commits/1ea42fbad838a2a0d0c91277547f88c26c4ee70f

On Fri, May 1, 2020 at 2:30 AM Christopher Fujino <notifications@github.com>
wrote:

> @peterlauri <https://github.com/peterlauri> I wanted to make a PR into
> your feature branch, but I couldn't figure out how to do it from the GitHub
> web UI. Can you patch this diff into your PR 5658f5e
> <https://github.com/flutter/flutter/commit/5658f5e42b301ed6fcd18bf118b75097a0df0962>?
> (If you fetch upstream, I bet you can cherry pick the exact commit, I
> committed on top of your branch.)
>
> It reverts the support for parsing the older version, and instead modifies
> the text that the flutter version tool prints out to not depend on
> parsing git tags.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/55602#issuecomment-622173351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA22EQIY4LU3O3WSTZLIUY3RPICX3ANCNFSM4MQNRDQA>
> .
>

christopherfujino commented:
> I applied that patch to the PR. [1ea42fb](https://github.com/flutter/flutter/pull/55602/commits/1ea42fbad838a2a0d0c91277547f88c26c4ee70f)
> […](#)
> On Fri, May 1, 2020 at 2:30 AM Christopher Fujino ***@***.***> wrote: @peterlauri <https://github.com/peterlauri> I wanted to make a PR into your feature branch, but I couldn't figure out how to do it from the GitHub web UI. Can you patch this diff into your PR [5658f5e](https://github.com/flutter/flutter/commit/5658f5e42b301ed6fcd18bf118b75097a0df0962) <[5658f5e](https://github.com/flutter/flutter/commit/5658f5e42b301ed6fcd18bf118b75097a0df0962)>? (If you fetch upstream, I bet you can cherry pick the exact commit, I committed on top of your branch.) It reverts the support for parsing the older version, and instead modifies the text that the flutter version tool prints out to not depend on parsing git tags. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#55602 (comment)](https://github.com/flutter/flutter/pull/55602#issuecomment-622173351)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA22EQIY4LU3O3WSTZLIUY3RPICX3ANCNFSM4MQNRDQA> .

Sorry, I had tests failing on that patch. I force pushed past it another version https://github.com/flutter/flutter/pull/56093/commits/46c1571c7397a687a156e0d8735f3a2f8e3d8531, which basically removes tests that were no longer necessary (or logical), since if the user was in a bad state the tool would exit earlier when we try to fetch tags. Can you try applying one more time?

peterlauri committed:
Revert "remove instantiating new FlutterVersion() in flutter version command. https://github.com/flutter/flutter/commit/5658f5e42b301ed6fcd18bf118b75097a0df0962"

This reverts commit 1ea42fbad838a2a0d0c91277547f88c26c4ee70f.

peterlauri committed:
remove instantiating new FlutterVersion() in `flutter version` command https://github.com/flutter/flutter/commit/46c1571c7397a687a156e0d8735f3a2f8e3d8531

peterlauri committed:
Merge remote-tracking branch 'upstream/master' into fix-flutter-version-dev-55383

peterlauri committed:
changed name of test to not have duplicate test names

peterlauri commented:
@christopherfujino reverted old patch and applied your patch. Also renamed one test as two tests had identical name "dev version switch prompt is accepted". Changed the old version to "old dev version switch prompt is accepted".

peterlauri commented:
Oh, and I merged upstream/master as well, just to be sure there were not changes upstream. Ran the tests locally, lets see what CI says...

christopherfujino reviewed (approved):
LGTM! Thanks again for this, I know this is not the easiest code to modify.
